cpp/include/raft/core/device_resources_snmg.hpp

```diff
@@ -48,193 +42,94 @@ class device_resources_snmg : public device_resources {
   device_resources_snmg() : device_resources()
   {
     RAFT_CUDA_TRY(cudaGetDevice(&main_gpu_id_));
+    raft::resource::set_root_rank(*this, 0);
 
-    // set root rank
-    int root_rank = 0;
-    raft::resource::set_nccl_clique_root_rank(*this, root_rank);
-
-    // initialize the NCCL clique
-    std::vector<raft::resources>& clique = raft::resource::get_nccl_clique(*this);
-    _init_clique(clique);
-    _init_nccl_comms(clique);
+    // initialize all resources
+    std::vector<raft::resources>& world_resources = raft::resource::get_multi_gpu_resource(*this);
+    _init_world(world_resources);
 
     RAFT_CUDA_TRY(cudaSetDevice(main_gpu_id_));
   }
 
   /**
    * @brief Construct a SNMG resources instance with a subset of available GPUs
    *
-   * @param[in] device_ids List of device IDs to be used by the NCCL clique
+   * @param[in] device_ids List of device IDs to be used
    */
   device_resources_snmg(const std::vector<int>& device_ids) : device_resources()
   {
     RAFT_CUDA_TRY(cudaGetDevice(&main_gpu_id_));
+    raft::resource::set_root_rank(*this, 0);
 
-    // set root rank
-    int root_rank = 0;
-    raft::resource::set_nccl_clique_root_rank(*this, root_rank);
-
-    // initialize the NCCL clique
-    std::vector<raft::resources>& clique = raft::resource::get_nccl_clique(*this);
-    _init_clique(clique, device_ids);
-    _init_nccl_comms(clique);
-
+    // initialize resources for the given device ids
+    std::vector<raft::resources>& world_resources = raft::resource::get_multi_gpu_resource(*this);
+    _init_world(world_resources, device_ids);
     RAFT_CUDA_TRY(cudaSetDevice(main_gpu_id_));
   }
 
   /**
    * @brief SNMG resources instance copy constructor
    *
-   * @param[in] clique A SNMG resources instance
+   * @param[in] world A SNMG resources instance
    */
-  device_resources_snmg(const device_resources_snmg& clique)
-    : device_resources(clique), main_gpu_id_(clique.main_gpu_id_)
-  {
-  }
+  device_resources_snmg(const device_resources_snmg& world) : device_resources(world) {}
```

Comment:
The copy constructor doesn't copy the main_gpu_id_, which causes [cudaErrorInvalidDevice: invalid device ordinal](https://github.com/rapidsai/cuvs/pull/869#issuecomment-2876235941) at a later point in time.
```suggestion
  device_resources_snmg(const device_resources_snmg& world)
    : device_resources(world), main_gpu_id_(world.main_gpu_id_)
  {
  }
```

Reply:
thank you for pointing this out, I was unable to test yesterday because of devcontainer issues.

Reply:
@achirkin As for the move constructor, I was sticking to the existing code, and it looks like other raft resources are set to be that way ([link](https://github.com/rapidsai/raft/blob/217189d67dab43cacb1b7f5076a6485502ef6010/cpp/include/raft/core/device_resources.hpp#L74-L76)). I'm thinking it's because we don't want to transfer ownership of resources. ccing @cjnolet for verification.

Reply:
Ah, right, that might be related to the mutex member in the parent resource. We had some thread-safety discussions related to that (each copy has its own mutex)